### PR TITLE
Fixes to update & watch functionality for TPRs

### LINKED ISSUE #406: Implement watch and watch-list functionality for TPR storage in the API server

### LINKED ISSUE #436: TPR-based API server cannot update status
The TPR based API server panics when PUT requests come in to update the status of a TPR. See below for the error in logs:

```console
E0228 17:49:30.788866       1 errors.go:63] apiserver received an error that is not an metav1.Status: not implemented for third party resources
I0228 17:49:30.789075       1 panics.go:76] PUT /apis/servicecatalog.k8s.io/v1alpha1/brokers/test-broker/status: (575.605µs) 500
goroutine 163 [running]:
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc4202e2e70, 0x1f4)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/httplog/log.go:219 +0xbb
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc4202e2e70, 0x1f4)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/httplog/log.go:198 +0x35
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/metrics.(*responseWriterDelegator).WriteHeader(0xc4202353e0, 0x1f4)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:117 +0x45
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver.writeNegotiated(0x22fbf20, 0xc42038b080, 0x18780ef, 0x15, 0x1864c03, 0x8, 0x22fabe0, 0xc4200240a0, 0xc4205305a0, 0x1f4, ...)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/apiserver.go:464 +0x162
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver.errorNegotiated(0x22e45e0, 0xc4203b5740, 0x22fbf20, 0xc42038b080, 0x18780ef, 0x15, 0x1864c03, 0x8, 0x22fabe0, 0xc4200240a0, ...)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/apiserver.go:481 +0x103
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver.(*RequestScope).err(0xc42032bdc0, 0x22e45e0, 0xc4203b5740, 0x22fabe0, 0xc4200240a0, 0xc4205305a0)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/resthandler.go:89 +0xd7
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver.UpdateResource.func1(0xc420235320, 0xc420791920)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/resthandler.go:727 +0xd63
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/metrics.InstrumentRouteFunc.func1(0xc420235320, 0xc420791920)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:101 +0x1f2
github.com/kubernetes-incubator/service-catalog/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc420371ef0, 0x22fad60, 0xc4202e2e70, 0xc4205305a0)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/github.com/emicklei/go-restful/container.go:272 +0xba0
github.com/kubernetes-incubator/service-catalog/vendor/github.com/emicklei/go-restful.(*Container).(github.com/kubernetes-incubator/service-catalog/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x22fad60, 0xc4202e2e70, 0xc4205305a0)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/github.com/emicklei/go-restful/container.go:120 +0x48
net/http.HandlerFunc.ServeHTTP(0xc42028c370, 0x22fad60, 0xc4202e2e70, 0xc4205305a0)
	/usr/local/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc420206300, 0x22fad60, 0xc4202e2e70, 0xc4205305a0)
	/usr/local/go/src/net/http/server.go:2022 +0x7f
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters.WithPanicRecovery.func1(0x22fad60, 0xc4202e2e70, 0xc4205305a0)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters/panics.go:75 +0x24a
net/http.HandlerFunc.ServeHTTP(0xc420206990, 0x7f5300eff820, 0xc420024090, 0xc4205305a0)
	/usr/local/go/src/net/http/server.go:1726 +0x44
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters.(*timeoutHandler).ServeHTTP.func1(0xc420442c00, 0x22fd0e0, 0xc420024090, 0xc4205305a0, 0xc420791800)
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters/timeout.go:89 +0x8d
created by github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters.(*timeoutHandler).ServeHTTP
	/Users/aaronschlesinger/gocode/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/pkg/genericapiserver/filters/timeout.go:91 +0x1db

logging error output: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"not implemented for third party resources\",\"code\":500}\n"
```

### PR DESCRIPTION
**Pull request by arschles**
This PR is a combination of #410 and #439. It will replace those two, but is not ready for merge yet.

Closes #410 
Closes #439 

Fixes #436 
Fixes #406

### CODE REVIEW

**Comment by arschles** [comment]
The functionality herein works properly. After creating a broker with an invalid URL, the controller properly picks it up and updates the status properly:

```console
ENG000656:service-catalog aaronschlesinger$ ks get broker -o yaml test-broker
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: Broker
metadata:
  creationTimestamp: 2017-03-06T18:53:47Z
  name: test-broker
  resourceVersion: "1493"
  selfLink: /apis/servicecatalog.k8s.io/v1alpha1/brokerstest-broker
  uid: 3ae44ce4-029e-11e7-a591-080027d16e09
spec:
  url: http://beefco.de
status:
  conditions:
  - message: Error fetching catalog
    reason: ErrorFetchingCatalog
    status: "False"
    type: Ready
```

File: pkg/storage/tpr/storage_versioner.go [hunk 1]
```diff
@@ -17,45 +17,92 @@ limitations under the License.
...
-	// }
-	// cl, err := GetResourceClient(t.cl, t.singularKind, namespace)
-	// if err != nil {
-	// 	return err
-	// }
-	// if _, err := cl.Update(unstruc); err != nil {
-	// 	return err
-	// }
+	if err := accessor.SetResourceVersion(obj, strconv.Itoa(int(resourceVersion))); err != nil {
+		glog.Errorf("setting resource version (%s)", err)
```
Comment (vaikas-google) [inline pkg/storage/tpr/storage_versioner.go]: Here and other error places, maybe just do the %v?

Comment (arschles) [inline pkg/storage/tpr/storage_versioner.go]: no need - errors conform to `fmt.Stringer`, so `%s` will print out a good, human-readable message

File: pkg/storage/tpr/obj_state.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+type errCouldntGetResourceVersion struct {
+	err error
+}
+
+func (e errCouldntGetResourceVersion) Error() string {
+	return fmt.Sprintf("couldn't get resource version (%s)", e.err)
+}
+
+type objState struct {
```
Comment (pmorie) [inline pkg/storage/tpr/obj_state.go]: This makes me wonder whether this code should exist upstream, possibly in `client-go`.  What do you think?  Is this generally useful for people working programmaticly with TPR?

Comment (arschles) [inline pkg/storage/tpr/obj_state.go]: It's certainly useful for people writing TPR `storage.Interface` implementations. I think that this whole suite of TPR-based `storage.Interface` stuff should be pushed upstream, and soon (possibly to `k8s.io/apiserver`)

File: pkg/storage/tpr/storage_versioner.go [hunk 3]
```diff
@@ -81,7 +128,7 @@ func (t *storageVersioner) UpdateList(obj runtime.Object, resourceVersion uint64
 	// if _, err := cl.Update(unstruc); err != nil {
 	// 	return err
 	// }
-	return nil
+	return errors.New("UpdateList not implemented")
```
Comment (arschles) [inline pkg/storage/tpr/storage_versioner.go]: #479 captures this

**Comment by pmorie** [comment]
Command decision on this one - we've seen it work and it's behind a feature flag.  I've verified that it works in my cluster and will merge on green.

**Pull request merged**